ansible_rulebook/action/pg_notify.py

```diff
@@ -0,0 +1,109 @@
+#  Copyright 2023 Red Hat, Inc.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import json
+import logging
+import uuid
+
+import xxhash
+from psycopg import AsyncClientCursor, AsyncConnection, OperationalError
+
+from .control import Control
+from .helper import FAILED_STATUS, Helper
+from .metadata import Metadata
+
+logger = logging.getLogger(__name__)
+
+MAX_MESSAGE_LENGTH = 7 * 1024
+MESSAGE_CHUNKED_UUID = "_message_chunked_uuid"
+MESSAGE_CHUNK_COUNT = "_message_chunk_count"
+MESSAGE_CHUNK_SEQUENCE = "_message_chunk_sequence"
+MESSAGE_CHUNK = "_chunk"
+MESSAGE_LENGTH = "_message_length"
+MESSAGE_XX_HASH = "_message_xx_hash"
+
+
+class PGNotify:
+    """The PGNotify action sends an event to a PG Pub Sub Channel
+    Needs
+    dsn https://www.postgresql.org/docs/current/libpq-connect.html
+    #LIBPQ-CONNSTRING-KEYWORD-VALUE
+    channel the channel name to send the notifies
+    event
+    """
+
+    def __init__(self, metadata: Metadata, control: Control, **action_args):
+        self.helper = Helper(metadata, control, "pg_notify")
+        self.action_args = action_args
+
+    async def __call__(self):
+        if not self.action_args["event"]:
+            return
+
+        try:
+            async with await AsyncConnection.connect(
```

Comment:
It might be beneficial to cache the connection between actions if they have the same connection info.  This could be stored as a dictionary on the module or class.

Reply:
We should measure the performance with and without this caching to see if there is a performance improvement.

Reply:
There is a separate package called psycopg[pool] which has an AsyncConnectionPool https://www.psycopg.org/psycopg3/docs/advanced/pool.html#connection-pools
We can test using the pooled connection in a different PR
```
async with AsyncConnectionPool(...) as pool:
    async with pool.connection() as conn:
        await conn.execute("SELECT something FROM somewhere ...")

        with conn.cursor() as cur:
            await cur.execute("SELECT something else...")
```